core/src/main/scala/org/apache/spark/rdd/util/PrefetchingIterator.scala

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.rdd.util
+
+import java.util.concurrent.atomic.AtomicReference
+import java.util.concurrent.locks.ReentrantLock
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.reflect.ClassTag
+import scala.util.Try
+
+import org.apache.spark.rdd.RDD
+
+/**
+ * Iterator that optionally prefetches next partition asynchronously
+ */
+private[spark] class PrefetchingIterator[T : ClassTag](
+    rdd: RDD[T],
+    prefetch: Boolean)
+  extends Iterator[Array[T]]
+    with Serializable {
+
+  private val partitionIterator = rdd.partitions.indices.iterator
+
+  private val lock = new ReentrantLock()
```

Comment:
Could you explain the need for a lock here?

Reply:
since job is executed in other thread, the need is it to thread-safely pass the job result back to iterator